diff --git a/CMakeLists.txt b/CMakeLists.txt
index a1293db..72ad83f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -760,7 +760,9 @@ if(WITH_TURBOJPEG OR WITH_TOOLS OR WITH_TESTS)
 endif()
 
 if(ENABLE_SHARED)
-  add_subdirectory(sharedlib)
+  if(WITH_TURBOJPEG)
+    add_subdirectory(sharedlib)
+  endif()
 endif()
 
 if(ENABLE_STATIC)
@@ -875,6 +877,16 @@ if(WITH_TURBOJPEG)
   endif()
 endif()
 
+set(PSJPEG_SOURCES ${JPEG_SOURCES} ${SIMD_TARGET_OBJECTS} ${SIMD_OBJS} psjpeg.c)
+set(CMAKE_C_VISIBILITY_PRESET hidden)
+add_library(psjpeg SHARED ${PSJPEG_SOURCES})
+target_include_directories(psjpeg PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/src)
+set_target_properties(psjpeg PROPERTIES DEFINE_SYMBOL DLLDEFINE)
+
+if(UNIX AND HAVE_VERSION_SCRIPT)
+  set_target_properties(psjpeg PROPERTIES LINK_FLAGS
+    "-Wl,--version-script='${CMAKE_CURRENT_SOURCE_DIR}/psjpeg.ver'")
+endif()
 if(WITH_TOOLS)
   if(WIN32)
     set(USE_SETMODE "-DUSE_SETMODE")
@@ -1986,6 +1998,18 @@ if(WITH_TURBOJPEG)
     DESTINATION ${CMAKE_INSTALL_INCLUDEDIR} COMPONENT include)
 endif()
 
+install(TARGETS psjpeg EXPORT ${CMAKE_PROJECT_NAME}Targets
+  INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+  ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+  LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+  RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
+if(MSVC AND CMAKE_C_LINKER_SUPPORTS_PDB)
+  install(FILES "$<TARGET_PDB_FILE:psjpeg>"
+    DESTINATION ${CMAKE_INSTALL_BINDIR} OPTIONAL)
+endif()
+install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/psjpeg.h ${CMAKE_CURRENT_BINARY_DIR}/jconfig.h
+  ${CMAKE_CURRENT_SOURCE_DIR}/src/jerror.h ${CMAKE_CURRENT_SOURCE_DIR}/src/jmorecfg.h
+  ${CMAKE_CURRENT_SOURCE_DIR}/src/jpeglib.h DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/psjpeg)
 if(ENABLE_STATIC)
   install(TARGETS jpeg-static EXPORT ${CMAKE_PROJECT_NAME}Targets
     INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
@@ -2046,9 +2070,9 @@ if(UNIX OR MINGW)
       DESTINATION ${CMAKE_INSTALL_MANDIR}/man1 COMPONENT man)
   endif()
 endif()
+if(WITH_TURBOJPEG)
 install(FILES ${CMAKE_CURRENT_BINARY_DIR}/pkgscripts/libjpeg.pc
   DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig COMPONENT lib)
-if(WITH_TURBOJPEG)
   install(FILES ${CMAKE_CURRENT_BINARY_DIR}/pkgscripts/libturbojpeg.pc
     DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig COMPONENT lib)
 endif()
