diff --git a/src/jconfig.h.in b/src/jconfig.h.in
index 6cb8296..046dac5 100644
--- a/src/jconfig.h.in
+++ b/src/jconfig.h.in
@@ -38,11 +38,8 @@
 
 #undef RIGHT_SHIFT_IS_UNSIGNED
 
-/* Define "boolean" as unsigned char, not int, per Windows custom */
-#ifndef __RPCNDR_H__            /* don't conflict if rpcndr.h already read */
-typedef unsigned char boolean;
-#endif
-#define HAVE_BOOLEAN            /* prevent jmorecfg.h from redefining it */
+#define NO_GETENV
+#define NO_PUTENV
 
 /* Define "INT32" as int, not long, per Windows custom */
 #if !(defined(_BASETSD_H_) || defined(_BASETSD_H))   /* don't conflict if basetsd.h already read */
diff --git a/src/jdapistd.c b/src/jdapistd.c
index 8f5f85a..6b66964 100644
--- a/src/jdapistd.c
+++ b/src/jdapistd.c
@@ -203,8 +203,8 @@ _jpeg_crop_scanline(j_decompress_ptr cinfo, JDIMENSION *xoffset,
   if (cinfo->master->lossless || cinfo->raw_data_out)
     ERREXIT(cinfo, JERR_NOTIMPL);
 
-  if ((cinfo->global_state != DSTATE_SCANNING &&
-       cinfo->global_state != DSTATE_BUFIMAGE) || cinfo->output_scanline != 0)
+  if ((cinfo->global_state < DSTATE_SCANNING ||
+       cinfo->global_state > DSTATE_BUFIMAGE) || cinfo->output_scanline != 0)
     ERREXIT1(cinfo, JERR_BAD_STATE, cinfo->global_state);
 
   if (!xoffset || !width)
diff --git a/src/jmorecfg.h b/src/jmorecfg.h
index a4df71c..74a479c 100644
--- a/src/jmorecfg.h
+++ b/src/jmorecfg.h
@@ -27,7 +27,7 @@
  * bytes of storage, whether actually used in an image or not.)
  */
 
-#define MAX_COMPONENTS  10      /* maximum number of image components */
+#define MAX_COMPONENTS   4      /* maximum number of image components */
 
 
 /*
@@ -236,8 +236,8 @@ typedef int boolean;
 /* Capability options common to encoder and decoder: */
 
 #define DCT_ISLOW_SUPPORTED     /* accurate integer method */
-#define DCT_IFAST_SUPPORTED     /* less accurate int method [legacy feature] */
-#define DCT_FLOAT_SUPPORTED     /* floating-point method [legacy feature] */
+
+
 
 /* Encoder capability options: */
 
@@ -245,7 +245,7 @@ typedef int boolean;
 #define C_PROGRESSIVE_SUPPORTED     /* Progressive JPEG?  (Requires
                                        C_MULTISCAN_FILES_SUPPORTED and
                                        ENTROPY_OPT_SUPPORTED) */
-#define C_LOSSLESS_SUPPORTED        /* Lossless JPEG? */
+
 #define ENTROPY_OPT_SUPPORTED       /* Optimization of entropy coding parms? */
 /* Note: if you selected 12-bit data precision, it is dangerous to turn off
  * ENTROPY_OPT_SUPPORTED.  The standard Huffman tables are only good for 8-bit
@@ -255,22 +255,20 @@ typedef int boolean;
  * The exact same statements apply for lossless JPEG: the default tables don't
  * work for lossless mode.  (This may get fixed, however.)
  */
-#define INPUT_SMOOTHING_SUPPORTED   /* Input image smoothing option? */
+
 
 /* Decoder capability options: */
 
 #define D_MULTISCAN_FILES_SUPPORTED /* Multiple-scan JPEG files? */
 #define D_PROGRESSIVE_SUPPORTED     /* Progressive JPEG?  (Requires
                                        D_MULTISCAN_FILES_SUPPORTED) */
-#define D_LOSSLESS_SUPPORTED        /* Lossless JPEG?  (Requires
-                                       D_MULTISCAN_FILES_SUPPORTED) */
 #define SAVE_MARKERS_SUPPORTED      /* jpeg_save_markers() needed? */
-#define BLOCK_SMOOTHING_SUPPORTED   /* Block smoothing? (Progressive only) */
+
 #define IDCT_SCALING_SUPPORTED      /* Output rescaling via IDCT?  (Requires
                                        DCT_ISLOW_SUPPORTED) */
-#define UPSAMPLE_MERGING_SUPPORTED  /* Fast path for sloppy upsampling? */
-#define QUANT_1PASS_SUPPORTED       /* 1-pass color quantization? */
-#define QUANT_2PASS_SUPPORTED       /* 2-pass color quantization? */
+
+
+
 
 /* more capability options later, no doubt */
 
