diff --git a/pngread.c b/pngread.c
--- a/pngread.c
+++ b/pngread.c
@@ -224,3 +224,3 @@ png_read_frame_head(png_structp png_ptr, png_infop info_ptr)
         /* Discard trailing fdATs for all frames except the first. */
-        if (!have_chunk_after_DAT && png_ptr->num_frames_read > 1)
+        if (!have_chunk_after_DAT && png_ptr->num_frames_read >= 1)
         {
@@ -527,5 +527,5 @@ png_read_row(png_structrp png_ptr, png_bytep row, png_bytep dsp_row)
    png_read_IDAT_data(png_ptr, png_ptr->row_buf, row_info.rowbytes + 1);
 
-   if (png_ptr->row_buf[0] > PNG_FILTER_VALUE_NONE)
+   if (png_ptr->row_buf[0] > PNG_FILTER_VALUE_NONE && png_ptr->row_buf[0] < PNG_FILTER_VALUE_LAST)
    {
       if (png_ptr->row_buf[0] < PNG_FILTER_VALUE_LAST)
diff --git a/pngwrite.c b/pngwrite.c
--- a/pngwrite.c
+++ b/pngwrite.c
@@ -1549,11 +1549,5 @@ png_write_frame_head(png_structp png_ptr, png_infop info_ptr,
    png_debug(1, "in png_write_frame_head");
 
-   /* There is a chance this has been set after png_write_info was called,
-    * so it would be set but not written. Is there a way to be sure?
-    */
-   if (!(info_ptr->valid & PNG_INFO_acTL))
-      png_error(png_ptr, "Cannot write APNG frame: missing acTL");
-
    png_write_reset(png_ptr);
 
    png_write_reinit(png_ptr, info_ptr, width, height);
diff --git a/pngwutil.c b/pngwutil.c
--- a/pngwutil.c
+++ b/pngwutil.c
@@ -2899,12 +2899,7 @@ png_write_reinit(png_structp png_ptr, png_infop info_ptr,
    if (width > png_ptr->first_frame_width ||
        height > png_ptr->first_frame_height)
       png_error(png_ptr, "Oversized frame in fcTL");
 
-   png_set_IHDR(png_ptr, info_ptr, width, height,
-                info_ptr->bit_depth, info_ptr->color_type,
-                info_ptr->interlace_type, info_ptr->compression_type,
-                info_ptr->filter_type);
-
    png_ptr->width = width;
    png_ptr->height = height;
    png_ptr->rowbytes = PNG_ROWBYTES(png_ptr->pixel_depth, width);
