diff --git a/libde265/CMakeLists.txt b/libde265/CMakeLists.txt
index d0d6762..06e1660 100644
--- a/libde265/CMakeLists.txt
+++ b/libde265/CMakeLists.txt
@@ -115,8 +115,8 @@ add_library(de265 ${libde265_sources} ${libde265_public_headers} ${ENCODER_OBJECTS} ${X86_OBJECTS} ${ARM32_OBJECTS})
 add_library(de265 ${libde265_sources} ${libde265_public_headers} ${ENCODER_OBJECTS} ${X86_OBJECTS} ${ARM32_OBJECTS})
-target_include_directories(de265 PRIVATE ${CMAKE_BINARY_DIR} ${CMAKE_CURRENT_BINARY_DIR})
+target_include_directories(de265 PRIVATE $<BUILD_INTERFACE:${CMAKE_BINARY_DIR} ${CMAKE_CURRENT_BINARY_DIR}> PUBLIC $<INSTALL_INTERFACE:include>)
 target_link_libraries(de265 PRIVATE Threads::Threads)
 if(UNIX)
   target_link_libraries(de265 PRIVATE m)
 endif()
 
 write_basic_package_version_file(libde265-config-version.cmake COMPATIBILITY ExactVersion)
